### DISCUSSION TITLE
Rename `global-imports` to `auto-import`

### DISCUSSION THREAD

**Discussion by antfu** [nuxt/nuxt#18360] [General]
Originally we made `global-imports` because of the `vue-global-api` packages. But since then we moved to the `on-demand` imports injecting approach, we no longer having them `globally` (while it's the case in dev, but I would think it's more like an optimization instead of making things really global) - so I am thinking maybe we we could change the internal module name to `auto-import` to align with `unplugin-auto-import` and been a bit more clear for what's it does.

Alternatives:

- `auto-import`
- `api-import`
- `auto-api-import`
- `auto-api`
- ...add yours

**Comment by pi0** [comment]
Makes sense to rename terminology. I would rather make it even closer to the use case like "script auto import" since this feature is only available within `<script setup>`.

**Reply by antfu** [reply]
Would it be better to have them also available outside of SFC? Since in dev the API is actually global available, and meanwhile, the type definition is also global. Enable auto importing for user-side scripts and SFC make more sense to me (which is what `unplugin-auto-import` is doing)

**Reply by pi0** [reply]
This might be tricky for performance as we have to apply parse/transform all ts/mjs files and potentially bailing some optimizations.

But supporting for non `.vue` makes sense (we was also talking about with @danielroe yesterday) that types are at least registering availability to `globalThis` that is typed on all files.
 
BTW Then these magic imports are actually globally available with an optimization for prod that smartly injects them by usage :D 

What about making it with other nuxt feature "Component Discovery" and call it "Import Discovery"? ("auto Import" is also fine for me)

**Reply by antfu** [reply]
Wow, love the Discovery series! 👍

**Reply by antfu** [reply]
While I was trying to make a PR for this change, I found that eventho we are branding components as "Compontent Discvoery", in implementation we are still calling it `components` in folders, modules names, etc. (which make sense to me). So I guess we should name this `imports` or `imports-discovery`?

The main reason for me to push this change is that I want to have a hook for modules to provide more identifiers so I can make auto import for VueUse out-of-box.

**Reply by antfu** [reply]
For the perf concern, I understand your points but there are also some points to reference:

- I see we are switching to Vite by default, and transforming overhead in Vite is relatively small I believe
- Auto importing does not parse the code, the regex match should be quite cheap and fast.
- When no imports matched, the transformation process is passed.
- It only applies to user code but not node_modules.
